### DISCUSSION TITLE
How can I stop Next.js from messing with my tsconfig?

### DISCUSSION THREAD

**Discussion by wereHamster** [vercel/next.js#58022] [Help]
### Summary

Whenever I start next dev server, it overwrites my tsconfig with its opinionated settings:

```
   We detected TypeScript in your project and reconfigured your tsconfig.json file for you. Strict-mode is set to false by default.
   The following suggested values were added to your tsconfig.json. These values can be changed to fit your project's needs:

   	- strictNullChecks was set to true
```

How can I stop Next.js from doing that?

**Comment by boydenhartog** [comment]
I'm also running into this issue with version: 13.5.6

**Comment by arimgibson** [comment]
I didn't have the issue with the `strictNullChecks` property but another value (`noEmit`) that said "These values can be changed" in the error message. It would default to adding it to true, and add it back in every time I deleted it. Instead, I kept the value there but set to `false` and the error stopped. Might help here!

**Comment by ssalbdivad** [comment]
This is genuinely terrible behavior if you don't want to allow certain settings just give a warning.

`node` is an outdated moduleResolution setting that doesn't respect `package.json` exports (added in Node 12 which is long since deprecated).

Why overwrite whatever I try to add with that? I've never seen a dev tool be that overly aggressive about screwing up my configs.

**Comment by adamshaylor** [comment]
I stumbled into this when I was importing some uncompiled React components from another library. I tried using the `Bundler` value. Rewriting my `tsconfig.json` was an unpleasant surprise. In some cases, it will even generate invalid combinations:

![Screenshot 2024-06-18 at 12 44 06](https://github.com/vercel/next.js/assets/673870/cd3c34aa-41b6-4c24-9aad-556235a133ef)

The flow of a build process should be unidirectional. A build tool that modifies source (and `tsconfig.json` is definitely source) is creating more problems than it solves.

**Comment by LuxXx** [comment]
I am using `extends`, that points to another tsconfig from another package (which also extends from a base tsconfig) using turborepo.
The base tsconfig is really strict, but 

`tsconfig.json`:
```json
{
  "extends": "@repo/typescript-config/nextjs.json",
  "compilerOptions": {
    "paths": {
      "@/*": ["./src/*"]
    },
  },
  "include": [
    "next-env.d.ts",
    "next.config.js",
    "**/*.ts",
    "**/*.tsx",
    ".next/types/**/*.ts"
  ],
  "exclude": ["node_modules"]
}
```
`nextjs.json`:
```json
{
  "$schema": "https://json.schemastore.org/tsconfig",
  "display": "Next.js",
  "extends": "./base.json",
  "compilerOptions": {
    "plugins": [{ "name": "next" }],
    "module": "ESNext",
    "moduleResolution": "Bundler",
    "allowJs": true,
    "jsx": "preserve",
    "noEmit": true
  }
}
```

`base.json`:

```json
{
  "$schema": "https://json.schemastore.org/tsconfig",
  "display": "Default",
  "compilerOptions": {
    "declaration": false,
    "declarationMap": false,
    "esModuleInterop": true,
    "incremental": true,
    "isolatedModules": true,
    "lib": ["es2022", "DOM", "DOM.Iterable"],
    "module": "NodeNext",
    "moduleDetection": "force",
    "moduleResolution": "NodeNext",
    "noUncheckedIndexedAccess": true,
    "resolveJsonModule": true,
    "skipLibCheck": true,
    "strict": true,
    "target": "ES2022",
    "forceConsistentCasingInFileNames": true
  }
}
```


Starting the dev server immediately overwrites my tsconfig even though it should be strict already.

```
We detected TypeScript in your project and reconfigured your tsconfig.json file for you. Strict-mode is set to false by default.
             │   The following suggested values were added to your tsconfig.json. These values can be changed to fit your project's needs:
             │
             │        - strictNullChecks was set to true
```

**Reply by ozdv** [reply]
Did you end up resolving this @LuxXx  ? I am encountering this now setting up tailwind with nextjs and turborepo as well, and nextjs overwriting it on build is causing issues

**Reply by vinnymac** [reply]
Here is how I worked around the issue above: https://gist.github.com/vinnymac/efa90137f8c7026490c4b19f8ab73286

In case it helps you.

**Comment by hbystrom91** [comment]
i have the same issue. why the heck is next messing with my tsconfig? due to legacy reasons, i don't want strictNullChecks to be set to true. this basically breaks my app.

**Comment by GoodbyeNJN** [comment]
I also struggled with this issue, and I don't think the vercel team will take the advice to leave tsconfig unchanged. So I wrote a little script dedicated to patch next.js, just simply run `npx stop-nextjs-from-messing-tsconfig` from the project folder, feel free to try it out.

Source code: https://www.npmjs.com/package/stop-nextjs-from-messing-tsconfig

**Comment by apetta** [comment]
The pursuit of optimal DX is a double edged sword - this ones hurts more than it helps!

**Comment by BrianEstrada** [comment]
this is still an issue, how do we fix this?

**Comment by CSharperMantle** [comment]
This is still a problem on Next.js 16.1.6.

This modification tries to include files from my [`distDir`](https://nextjs.org/docs/app/api-reference/config/next-config-js/distDir) -- the artifact directory -- into the source config. Any sound build tool will *not* make such a decision. So sad to see this still being an issue after three years.

```console
$ pnpm dev
[...]
▲ Next.js 16.1.6 (Turbopack)
[...]

  We detected TypeScript in your project and reconfigured your tsconfig.json file for you.
  The following suggested values were added to your tsconfig.json. These values can be changed to fit your project's needs:

        - include was updated to add 'out/types/**/*.ts'
        - include was updated to add 'out/dev/types/**/*.ts'
```

**Comment by Javimtib92** [comment]
[Typescript 6.x strict defaults to true ](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-6-0.html#simple-default-changes)which is the option I want to use so I removed that configuration to clean up a bit the `tsconfig` file and NextJS keeps adding `strict: false`.


 > We detected TypeScript in your project and reconfigured your tsconfig.json file for you. Strict-mode is set to false by default.
  The following suggested values were added to your tsconfig.json. These values can be changed to fit your project's needs:

        - strict was set to false
        
 There's another discussion thread started in 2022 similar to this https://github.com/vercel/next.js/discussions/39942
 
 I feel like community is requesting for a way to opt out to this and we are not getting heard.

**Reply by paulusesterhazy-nelly** [reply]
There's no reason a build command should change tsconfig.json - whoever thought this was a good idea in the first place?
